chain/chunks/src/lib.rs

```diff
@@ -1018,228 +1018,224 @@ impl ShardsManager {
     ) -> (std::time::Instant, &'static str, Option<PartialEncodedChunkResponseMsg>) {
         // Try getting data from in-memory cache.
         let started = Instant::now();
-        if let Some(entry) = self.encoded_chunks.get(&request.chunk_hash) {
-            let response = Self::prepare_partial_encoded_chunk_response_from_cache(request, entry);
-            return (started, "cache", response);
+        let PartialEncodedChunkRequestMsg { chunk_hash, part_ords, mut tracking_shards } = request;
+        let mut part_ords = part_ords.into_iter().collect::<HashSet<_>>();
+        if part_ords.is_empty() && tracking_shards.is_empty() {
+            return (started, "none", None);
+        }
+        let mut response = PartialEncodedChunkResponseMsg {
+            chunk_hash: chunk_hash.clone(),
+            parts: vec![],
+            receipts: vec![],
+        };
+        if let Some(entry) = self.encoded_chunks.get(&chunk_hash) {
+            Self::lookup_partial_encoded_chunk_from_cache(
+                &mut part_ords,
+                &mut tracking_shards,
+                &mut response,
+                entry,
+            );
+        }
+        if part_ords.is_empty() && tracking_shards.is_empty() {
+            return (started, "cache", Some(response));
         }
 
         // Try fetching partial encoded chunk from storage.
-        let started = Instant::now();
-        if let Ok(partial_chunk) = self.store.get_partial_chunk(&request.chunk_hash) {
-            let response =
-                Self::prepare_partial_encoded_chunk_response_from_partial(request, &partial_chunk);
-            return (started, "partial", response);
+        if let Ok(partial_chunk) = self.store.get_partial_chunk(&chunk_hash) {
+            Self::lookup_partial_encoded_chunk_from_partial_chunk_storage(
+                &mut part_ords,
+                &mut tracking_shards,
+                &mut response,
+                &partial_chunk,
+            );
+        }
+        if part_ords.is_empty() && tracking_shards.is_empty() {
+            return (started, "partial", Some(response));
         }
 
         // Try fetching chunk from storage and recomputing encoded chunk from
         // it.  If we are archival node we might have garbage collected the
         // partial chunk while we still keep the chunk itself.  We can get the
         // chunk, recalculate the parts and respond to the request.
-        let started = Instant::now();
-        if let Ok(chunk) = self.store.get_chunk(&request.chunk_hash) {
-            let response = self.prepare_partial_encoded_chunk_response_from_chunk(request, &chunk);
-            return (started, "chunk", response);
+        if let Ok(chunk) = self.store.get_chunk(&chunk_hash) {
+            self.lookup_partial_encoded_chunk_from_chunk_storage(
```

Comment:
Actually, this additional lookup from column Chunk should only be performed on archival nodes if the PartialEncodedChunk column is gc'ed. I think we always ensure that if the chunk is reconstructed, then all parts will be stored in PartialEncodedChunk column, so if the chunk exists here, there is no additional need to look from column Chunk again, so I think we should keep the original behavior here.

Reply:
I think that part of logic is correct. If we track the shard, then we persist all parts and receipts in PartialEncodedChunk, which is correct. If we don't track the shard, then we only persist the part we own and the receipt to the shards we care about, which is already correct. So back to the logic here. If we track the shard, then all parts will be persisted in PartialEncodedChunk and there is no need to look more in Chunk. If we don't track the shard, there is no need to look in Chunk either.

Reply:
Ah OK, this is interesting. I think some of the logic around this is inconsistent:
 - We only persist a part in PartialEncodedChunk if we care about the part, but we still keep the part in the in-memory cache even if we don't care about it. Should we change it so that the in-memory cache also doesn't persist the part (and actually, don't even bother processing such a part? It can save unnecessary CPU). Seems like this is covered in #3662.
 -  If we happen to have received enough parts (either by receiving more parts than needed due to the above mechanism, or by owning too many parts in some weird corner case) to reconstruct the chunk, we will reconstruct the whole chunk and persist the ShardChunk even if we don't care about this shard. Should we avoid reconstructing the full chunk (and avoid persisting ShardChunk) if we don't care about the shard?

If both of these are taken care of, then the logic can just be:
 - Filter out all parts/receipts we don't care about (if we can't even filter, it means we don't know the prev_hash, so return empty)
 - Look in cache
 - Look on disk for any remaining parts/receipts; if this is an old chunk then look in ShardChunk; otherwise look in PartialEncodedChunk. (I don't like the dependency on GC though... I wonder if the logic should just be, look in partial, if not found at all, then look in chunk (and only do that for archival nodes)).

Reply:
Good observation! I also don't like the dependency on GC. The alternative you suggested makes sense.

Reply:
Regarding #3662, this is actually not possible to do perfectly in today's implementation because if the chunk is an orphan we don't know if we need the part or receipt. We can only do this in non-orphan cases, which still works for optimizing CPU but does not affect the discussion here.

However, having extra entries in the in-memory cache seems to be fine since we look up the cache first. So we should be fine as long as we fix the second issue that we won't persist the complete ShardChunk even if we can reconstruct. I've included that in this revision.

Also minor tweak that I'm no longer returning an Option. It seems nice to just send back a message saying, "I can't find the parts you're looking for for this hash", and it makes the code simpler. I don't think it would add to a significant amount of networking messaging because it should be rare that these requests miss, and this message is so tiny and takes no effort to process.

Reply:
What about "Filter out all parts/receipts we don't care about"? I can't seem to find that. Am I missing something?

Reply:
That's the comment about #3662; we can't always filter this because the chunk may be an orphan and we don't know whether we care about a part/receipt. And if we can't do it consistently then we may as well not do it, or at least it isn't compelling to do it in this context.

Reply:
Should we filter the parts/receipts after we check the cache but before we check storage? Since we only persist parts/receipts we are about.

Reply:
I see. Yeah, good point. Let me at least tackle this in a separate PR since this fixes Nayduck tests. I'm not exactly sure if the extra complexity is worth the savings on disk access, should probably look at some metrics?